### PR TITLE
Rate limiter refactor test

### PR DESCRIPTION
**Pull request by vladis84**
| Q             | A
| ------------- | ---
| Is bugfix?    | no
| New feature?  |no
| Breaks BC?    |no
| Tests pass?   | yes
| Fixed issues  | comma-separated list of tickets # fixed by the PR, if any

### CODE REVIEW

File: framework/filters/RateLimiter.php [hunk 1]
```diff
@@ -61,26 +61,35 @@ class RateLimiter extends ActionFilter
...
+    }
 
     /**
      * @inheritdoc
      */
     public function beforeAction($action)
     {
-        $user = $this->user ? : (Yii::$app->getUser() ? Yii::$app->getUser()->getIdentity(false) : null);
-        if ($user instanceof RateLimitInterface) {
+        if (!$this->user && Yii::$app->getUser()) {
```
Comment (SilverFire) [inline framework/filters/RateLimiter.php]: I think `!$this->user` could be replaced to `$this->user === null`

**Review by SilverFire — CHANGES_REQUESTED** [review]
The tests seems to be broken: https://travis-ci.org/yiisoft/yii2/jobs/222293533

**Comment by vladis84** [comment]
I correct tests

**Comment by vladis84** [comment]
I corrected tests

File: framework/CHANGELOG.md [hunk 2]
```diff
@@ -3,7 +3,7 @@ Yii Framework 2 Change Log
 
 2.0.12 under development
 --------------------------
-
+- Enh: Test write for yii\filters\RateLimiter and refactoring (vladis84)
```
Comment (samdark) [inline framework/CHANGELOG.md]: `Enh #13994: Refactored yii\filters\RateLimiter. Added tests (vladis84)`

File: framework/filters/RateLimiter.php [hunk 3]
```diff
@@ -61,26 +61,35 @@ class RateLimiter extends ActionFilter
      */
     public $response;
 
+    public function init()
```
Comment (samdark) [inline framework/filters/RateLimiter.php]: phpdoc missing

File: framework/filters/RateLimiter.php [hunk 4]
```diff
@@ -61,26 +61,35 @@ class RateLimiter extends ActionFilter
...
+    {
+        if (!$this->request) {
```
Comment (samdark) [inline framework/filters/RateLimiter.php]: What are these initializations for?

File: tests/framework/filters/stubs/UserIdentity.php [hunk 5]
```diff
@@ -63,4 +64,19 @@ public function validateAuthKey($authKey)
     {
         throw new NotSupportedException();
     }
+
+    public function getRateLimit($request, $action)
```
Comment (samdark) [inline tests/framework/filters/stubs/UserIdentity.php]: Doesn't look like a good idea to me. Interface should return an array in this case but it does not. Implementing methods like that isn't right. Better to leave default user implementation w/o this interface at all.

File: tests/framework/filters/stubs/UserIdentity.php [hunk 6]
```diff
@@ -4,14 +4,15 @@
...
+use yii\filters\RateLimitInterface;
 use yii\web\IdentityInterface;
 
 /**
  * Class UserIdentity
  * @author Dmitry Naumenko <d.naumenko.a@gmail.com>
  * @since 2.0.7
  */
-class UserIdentity extends Component implements IdentityInterface
+class UserIdentity extends Component implements IdentityInterface, RateLimitInterface
```
Comment (samdark) [inline tests/framework/filters/stubs/UserIdentity.php]: Doesn't look right.

**Review by samdark — CHANGES_REQUESTED** [review]

File: tests/framework/filters/stubs/UserIdentity.php (hunk 5 above)
Comment (vladis84) [inline tests/framework/filters/stubs/UserIdentity.php]: In this case, the api will be broken.
The programmer must implement the RatLimitInterface for the use of RateLimiter.
The structure is reliable than the array.

File: framework/filters/RateLimiter.php (hunk 4 above)
Comment (vladis84) [inline framework/filters/RateLimiter.php]: https://github.com/yiisoft/yii2/blob/master/framework/filters/RateLimiter.php#L75
It is difficult to say why in the filter these properties

File: tests/framework/filters/stubs/UserIdentity.php (hunk 5 above)
Comment (samdark) [inline tests/framework/filters/stubs/UserIdentity.php]: Won't it be broken in case of returning nulls as well?

Comment (vladis84) [inline tests/framework/filters/stubs/UserIdentity.php]: no https://github.com/yiisoft/yii2/pull/13994/files#diff-f4d94127b38f18704a417ac109ef0af7R113

File: framework/filters/RateLimiter.php (hunk 3 above)
Comment (vladis84) [inline framework/filters/RateLimiter.php]: i corrected

File: tests/framework/filters/stubs/UserIdentity.php (hunk 6 above)
Comment (vladis84) [inline tests/framework/filters/stubs/UserIdentity.php]: I corrected

File: tests/framework/filters/stubs/UserIdentity.php (hunk 5 above)
Comment (vladis84) [inline tests/framework/filters/stubs/UserIdentity.php]: I corrected

File: framework/filters/RateLimiter.php [hunk 7]
```diff
@@ -61,26 +61,38 @@ class RateLimiter extends ActionFilter
      */
     public $response;
 
+    /**
+     * {inheritdoc}
+     */
+    public function init()
+    {
+        if (!$this->request) {
```
Comment (cebe) [inline framework/filters/RateLimiter.php]: test for `===` null instead of `!`.

File: framework/filters/RateLimiter.php [hunk 8]
```diff
@@ -61,26 +61,38 @@ class RateLimiter extends ActionFilter
...
+           $this->request = Yii::$app->getRequest();
+        }
+
+        if (!$this->response) {
```
Comment (cebe) [inline framework/filters/RateLimiter.php]: test for `===` null instead of `!`.

**Review by cebe — CHANGES_REQUESTED** [review]

**Review by samdark — APPROVED** [review]

**Review by cebe — APPROVED** [review]

**Review by SilverFire — APPROVED** [review]

**Comment by SilverFire** [comment]
Thank you!

**Pull request merged**